File: Data Structures & Algorithms/max-area-of-island/submission-1.java
class Solution {
    public int dfs(int i, int j, boolean vis[][], int[][] grid, int n, int m){
        if(i<0 || j<0 || i>= m || j>=n || vis[i][j] || grid[i][j] != 1) return 0;

        vis[i][j] = true;
        int area = 1;

        area += dfs(i-1,j, vis, grid, n, m);
        area += dfs(i,j+1, vis, grid, n, m);
        area += dfs(i+1, j, vis, grid, n, m);
        area += dfs(i,j-1, vis, grid, n, m);
        return area;
    }


    public int maxAreaOfIsland(int[][] grid) {
        int maxarea = 0;
        int m = grid.length;
        int n = grid[0].length;

        boolean vis[][] = new boolean[m][n];

        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j] == 1 && vis[i][j] == false ){
                    int area = dfs(i,j,vis, grid, n, m);
                    maxarea = Math.max(maxarea, area);
                }
            }
        }
        return maxarea;
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-1.cpp
class Solution {
public:
    void dfs(int i, int j, vector<vector<bool>>& vis, vector<vector<char>>& grid, int n, int m ){
        if( i<0 || j<0 || i>= m || j>= n || vis[i][j]== true || grid[i][j] != '1' ){
            return;
        }

        vis[i][j] = true;

        dfs(i-1, j, vis,grid, n, m);
        dfs(i, j+1, vis, grid, n, m);
        dfs(i+1, j, vis, grid, n, m);
        dfs(i,j-1, vis, grid, n, m);
    }


    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int islands = 0;

        // bool vis[n][m] = false;
        vector<vector<bool>> vis(m, vector<bool>(n, false));

        for(int i = 0; i<m; i++){
            for (int j = 0; j<n; j++){
                if(grid[i][j] == '1' && vis[i][j] == false){
                    dfs(i,j, vis, grid, n, m);
                    islands++;                }
            }

        }
        return islands;

        
    }
};


